examples/channelize_poly_bench.cu

```diff
@@ -0,0 +1,143 @@
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+/////////////////////////////////////////////////////////////////////////////////
+
+#include "matx.h"
+#include <cassert>
+#include <cstdio>
+#include <cmath>
+#include <memory>
+#include <fstream>
+#include <istream>
+#include <cuda/std/complex>
+
+using namespace matx;
+
+// This example is used primarily for development purposes to benchmark the performance of the
+// polyphase channelizer kernel(s). Typically, the parameters below (batch size, filter
+// length, input signal length, and channel range) will be adjusted to a range of interest
+// and the benchmark will be run with and without the proposed kernel changes.
+
+constexpr int NUM_WARMUP_ITERATIONS = 2;
+
+// Number of iterations per timed test. Iteration times are averaged in the report.
+constexpr int NUM_ITERATIONS = 20;
+
+template <typename InType, typename OutType>
+void ChannelizePolyBench(matx::index_t channel_start, matx::index_t channel_stop)
+{
+  struct {
+    matx::index_t num_batches;
+    matx::index_t filter_len_per_channel;
+    matx::index_t input_len;
+  } test_cases[] = {
+    { 1, 17, 256 },
+    { 1, 17, 3000 },
+    { 1, 17, 31000 },
+    { 1, 17, 256000 },
+    { 42, 17, 256000 },
+    { 128, 17, 256000 },
+  };
+
+  cudaStream_t stream;
+  cudaStreamCreate(&stream);
+  cudaEvent_t start, stop;
+  cudaEventCreate(&start);
+  cudaEventCreate(&stop);
+
+
+  for (size_t i = 0; i < sizeof(test_cases)/sizeof(test_cases[0]); i++) {
+    for (matx::index_t num_channels = channel_start; num_channels <= channel_stop; num_channels++) {
+      const matx::index_t num_batches = test_cases[i].num_batches;
+      const matx::index_t filter_len = test_cases[i].filter_len_per_channel * num_channels;
+      const matx::index_t input_len = test_cases[i].input_len;
+      const matx::index_t output_len_per_channel = (input_len + num_channels - 1) / num_channels;
+
+      auto input = matx::make_tensor<InType, 2>({num_batches, input_len});
+      auto filter = matx::make_tensor<InType, 1>({filter_len});
+      auto output = matx::make_tensor<OutType, 3>({num_batches, output_len_per_channel, num_channels});
+
+      input.PrefetchDevice(stream);
+      filter.PrefetchDevice(stream);
+      output.PrefetchDevice(stream);
+      cudaStreamSynchronize(stream);
+
+      const matx::index_t decimation_factor = num_channels;
+
+      for (int k = 0; k < NUM_WARMUP_ITERATIONS; k++) {
+        (output = channelize_poly(input, filter, num_channels, decimation_factor)).run(stream);
+      }
+
+      cudaStreamSynchronize(stream);
+
+      double total_elapsed_ms = 0.0;
+      float iter_elapsed_ms = 0.0f;
+      for (int k = 0; k < NUM_ITERATIONS; k++) {
+        cudaEventRecord(start);
```

Comment:
Would it be better to have the events/sync outside the loop?

Reply:
Yep - I lifted the events out of the loop and removed the prefetch in an update.